processors/conversion/extract_urls.py

```diff
@@ -0,0 +1,347 @@
+
+        self.dataset.finish(url_matches_found)
+
+    @staticmethod
+    def resolve_redirect(url, redirect_domains=None, cache={}, depth=0):
+        """
+        Attempt to resolve redirects
+
+        :param str url: URL to check for redirect
+        :param tuple redirect_domains: Tuple with all domains to check for redirects
+        :param dict cache: URL cache updated with original cache[original_url] = redirect_url
+        :param int depth: Number of redirects to attempt to follow
+        :return str: Original url or new url for redirects
+        """
+        # Can use regex.sub() instead of string
+        if hasattr(url, "group"):
+            url = url.group(0)
+
+        # get host name to compare to list of shorteners
+        host_name = re.sub(r"^[a-z]*://", "", url).split("/")[0].lower()
+
+        if depth >= 10:
+            return url
+
+        elif url in cache:
+            return cache[url]
+
+        elif redirect_domains is None:
+            # No redirect_domains passed; check for redirect!
+            pass
+
+        elif "api.parler.com/l" not in url and host_name not in redirect_domains:
+            # skip non-redirects
+            return url
+
+
+        # to avoid infinite recursion, do not go deeper than 5 loops and
+        # keep track of current depth here:
+        depth += 1
+
+        # do this explicitly because it is a known issue and will save
+        # one request
+        if host_name == "t.co" and "http://" in url:
+            url = url.replace("http://", "https://")
+
+        try:
+            time.sleep(0.1)
+            head_request = requests.head(url, timeout=5)
+        except (requests.RequestException, ConnectionError, ValueError, TimeoutError) as e:
+            return url
+
+        # if the returned page's status code is in the 'valid request'
+        # range, and if it has a Location header different from the page's
+        # url, recursively resolve the page it redirects to up to a given
+        # depth - infinite recursion is prevented by using a cache
+        if 200 <= head_request.status_code < 400:
+            redirected_to = head_request.headers.get("Location", url)
+            if redirected_to != url:
+                cache[url] = redirected_to
+                return ExtractURLs.resolve_redirect(redirected_to, redirect_domains, cache, depth)
+
+        return url
+
+    @staticmethod
+    def identify_links(text):
+        """
+        Search string of text for URLs that may contain links.
+
+        :param str text:  string that may contain URLs
+        :return list:  	  list containing validated URLs to videos
+        """
+        # Extracting all links
+        # From https://stackoverflow.com/questions/6038061/regular-expression-to-find-urls-within-a-string
+        link_regex = re.compile(r"(http|ftp|https):\/\/([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:\/~+#-]*[\w@?^=%&\/~+#-])", re.IGNORECASE)
```

Comment:
Did you also try [ural](https://github.com/medialab/ural#urls_from_text) for this? To outsource maintenance of the regex ;)

Reply:
Nope, but that sounds great. I'll make that change.